File: LeetCode_Algorithm/1920(Build Array from Permutation)/1920.js
var buildArray = function(nums) {
  // sol1 88ms 61% 46.1MB 43%
  // let ans = []
  // for (let i=0; i < nums.length; i++) {
  //   ans.push(nums[nums[i]])
  // }
  // return ans

  // sol2 89ms 59% 45.7MB 69%
  for (let i=0; i < nums.length; i++) {
    if (nums[i] < nums.length) {
      nums[i] = nums[i] + nums.length * (nums[nums[i]] % nums.length)
    }
  }
  for (let i = 0; i < nums.length; i++) {
    nums[i] = parseInt(nums[i] / nums.length)
  }
  return nums
};

File: LeetCode_Algorithm/1920(Build Array from Permutation)/1920.py
'''
# sol1 129ms 65%
O(n) O(n)

# sol2 245ms 31% 14.1MB 86%
O(n) O(1)
해당 숫자에 접근하여 길이의 곱으로 만든 뒤
몫을 구한다.

'''
class Solution:
    def buildArray(self, nums: list[int]) -> list[int]:
        # sol1
        # ans = []
        # for i in range(len(nums)):
        #     ans.append(nums[nums[i]])
        # return ans

        # sol2
        for i in range(len(nums)):
            if nums[i] < len(nums):
                nums[i] = nums[i] + len(nums) * (nums[nums[i]] % len(nums))
        for i in range(len(nums)):
            nums[i] //= len(nums)
        return nums